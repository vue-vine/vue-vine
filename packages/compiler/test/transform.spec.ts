import { describe, expect, test } from 'vitest'
import type { VineCompilerHooks } from '../index'
import { compileVineTypeScriptFile, createCompilerCtx } from '../index'

function createMockTransformCtx(option = {}) {
  const mockCompilerCtx = createCompilerCtx(option)
  const mockCompilerHook = {
    onOptionsResolved: cb => cb(mockCompilerCtx.options),
    onError: () => {},
    onWarn: () => {},
    onBindFileCtx: (fileId, fileCtx) => mockCompilerCtx.fileCtxMap.set(fileId, fileCtx),
  } as VineCompilerHooks

  return {
    mockCompilerHook,
    mockCompilerCtx,
  }
}

describe('transform vcf containing await', () => {
  test('do not need result', () => {
    const content = 'export async function App() {\n'
      + '  const p = () => new Promise(resolve => setTimeout(resolve))\n'
      + '  await p()\n'
      + '  return vine`\n'
      + '    <div>test</div>\n'
      + '  `\n'
      + '}\n'
    const { mockCompilerHook } = createMockTransformCtx()
    const res = compileVineTypeScriptFile(content, 'testVCFAwait', mockCompilerHook)
    const code = res.fileSourceCode.toString()
    const matchRes = code.match(/await/g)
    expect(matchRes).toBeTruthy()
    expect(matchRes!.length).toBe(1)
    expect(code).toMatchSnapshot()
  })

  test('need result', () => {
    const content = 'export async function App() {\n'
      + '  const p = () => new Promise(resolve => setTimeout(resolve))\n'
      + '  const res = await p()\n'
      + '  return vine`\n'
      + '    <div>test</div>\n'
      + '  `\n'
      + '}\n'
    const { mockCompilerHook } = createMockTransformCtx()
    const res = compileVineTypeScriptFile(content, 'testVCFAwait', mockCompilerHook)
    const code = res.fileSourceCode.toString()
    const matchRes = code.match(/await/g)
    expect(matchRes).toBeTruthy()
    expect(matchRes!.length).toBe(1)
    expect(code).toMatchSnapshot()
  })
})

describe('transform vcf containing valid top level declaration', () => {
  test('should know top level declared names as LITERAL_CONST', () => {
    const content = 'const foo = \'foo\'\n'
    + 'const bar = () => \'lorem\'\n'
    + 'export function App() {\n'
    + '  console.log(foo, bar())\n'
    + '  return vine`\n'
    + '    <div>{{foo}} {{bar()}}</div>\n'
    + '  `\n'
    + '}'
    const { mockCompilerHook } = createMockTransformCtx()
    const res = compileVineTypeScriptFile(content, 'testVCFContainsTopLevelDecl', mockCompilerHook)
    const code = res.fileSourceCode.toString()
    expect(code).toMatchSnapshot()
  })
})
