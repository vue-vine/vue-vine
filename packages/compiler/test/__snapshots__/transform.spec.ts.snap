// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform vcf containing await > do not need result 1`] = `
"import {
  openBlock as _openBlock,
  createElementBlock as _createElementBlock,
  defineComponent as _defineComponent
} from 'vue'



export const App = (() => {

/* No hoisted */

  const __vine = _defineComponent({
    name: 'App',
    /* No props */
    /* No emits */
    async setup(__props /* No setup ctx destructuring */) {

const props = __props




const p = () => new Promise(resolve => setTimeout(resolve))
;(
  ([__temp,__restore] = _withAsyncContext(() => p())),
  await __temp,
  __restore()
)

/* No expose */

return (_ctx: any,_cache: any) => {
  return (_openBlock(), _createElementBlock(\\"div\\", null, \\"test\\"))
}

    }, // End of setup function
  }) // End of component object



  
  return __vine /* End of App */
})()"
`;

exports[`transform vcf containing await > need result 1`] = `
"import {
  openBlock as _openBlock,
  createElementBlock as _createElementBlock,
  defineComponent as _defineComponent
} from 'vue'



export const App = (() => {

/* No hoisted */

  const __vine = _defineComponent({
    name: 'App',
    /* No props */
    /* No emits */
    async setup(__props /* No setup ctx destructuring */) {

const props = __props




const p = () => new Promise(resolve => setTimeout(resolve))
const res = (
  ([__temp,__restore] = _withAsyncContext(() => p())),
  __temp = await __temp,
  __restore(),
  __temp
)

/* No expose */

return (_ctx: any,_cache: any) => {
  return (_openBlock(), _createElementBlock(\\"div\\", null, \\"test\\"))
}

    }, // End of setup function
  }) // End of component object



  
  return __vine /* End of App */
})()"
`;
